### PR TITLE
[ROCm] Enabling device tracing/profling support for AMD GPUs

### PR DESCRIPTION
**Pull request by deven-amd**
/cc @cheshire @chsigg

### CODE REVIEW

**Comment by deven-amd** [comment]
@cheshire @chsigg gentle ping

**Comment by deven-amd** [comment]
@cheshire @chsigg @gbaned gentle ping

**Comment by deven-amd** [comment]
@cheshire @chsigg @gbaned gentle ping

**Comment by deven-amd** [comment]
@kkimdev @rohan100jain @cheshire @chsigg @gbaned  gentle ping

**Comment by deven-amd** [comment]
@kkimdev @rohan100jain @cheshire @chsigg @gbaned gentle ping

the goal is to have this PR merged before we branch for TF 2.5, and hence the urgency .... please review.

**Comment by kkimdev** [comment]
Updated reviewers based on the files changed.  Reviewers: please redirect if needed.

**Comment by deven-amd** [comment]
@jbaiocchi @timshen91 gentle ping

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (timshen91) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h:37]: Please don't use `__` initiated names, they are reserved by the standard.

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (timshen91) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h:64]: ditto.

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (timshen91) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h:33]: Per style guide, don't create namespace aliases in headers.

**Review by timshen91 — COMMENTED** [review]
I'm only reviewing stream_executor.

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h]: changes `__name` to `xxx_name`

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h]: changes `__name` to `xxx_name`

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h]: removed

**Comment by deven-amd** [comment]
@timshen91 please re-review

**Comment by deven-amd** [comment]
@cheshire @chsigg @jbaiocchi gentle ping

**Comment by deven-amd** [comment]
@yisitu thanks for the response.

I appreciate the heads-up on the forth coming RFC...can you provide me a pointer to it.

In the meantime can we review and approve this quicker? I really want this functionality in TF 2.5, and would like to get this functionality merged in `master` so that it can be cherry-picked into the `r2.5` branch.

Most (all if you exclude the file renaming for cuda) are ROCm specific and should not have any impact on existing functionality.

w.r.t the new profiler add-on, we will be more than happy to update our implementation to use it, when it becomes available.

thanks

File: tensorflow/core/profiler/internal/gpu/BUILD
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/BUILD:147]:
```suggestion
    copts = tf_profiler_copts() + tf_copts(),
```

Comment (yisitu) [reply tensorflow/core/profiler/internal/gpu/BUILD:147]: Just to keep this consistent with the above.

File: tensorflow/core/profiler/internal/gpu/device_tracer_test.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/device_tracer_test.cc:269]: Instead of adding macros within the test (inline macros hurts readability), please duplicate this and have a specialized test for ROCM thath is guarded by TENSORFLOW_USE_ROCM.

File: tensorflow/core/profiler/internal/gpu/BUILD
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/BUILD:147]: done

File: tensorflow/core/profiler/internal/gpu/device_tracer_test.cc
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/device_tracer_test.cc]: done

**Comment by deven-amd** [comment]
@yisitu I have pushed out a commit with the changes you requested. please approve. thanks

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:37]:
```suggestion
  uint32_t destination;
```

Comment (yisitu) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:37]: (Please fix here and others below) we are using standard uint types going forward.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:55]: These can be uint32_t right? I don't think their sizes can be >4GB. It will make this compatible with cupti_collector.h::KernelDetails and make it easier for refactoring later on.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:141]: Does this need to be an ordered container? If not, please use absl::flat_hash_map for better performance.

Same for activity_tracing.

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:1]:
```suggestion
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:1]:
```suggestion
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (yisitu) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h:1]:
```suggestion
/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
```

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (yisitu) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h:65]: nit: might be night to align these backslashes vertically.

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (yisitu) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h:60]:
```suggestion
#endif  // PLATFORM_GOOGLE
```

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:289]: Use absl::flat_hash_set?

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:180]: We actually discourage TF_DISALLOW_COPY_AND_ASSIGN for readability reasons. For this and below, could you move the deleted interface within public accessor scope?

i.e.
```suggestion
 public:
  // Not copyable or movable
  AnnotationMap(const AnnotationMap&) = delete;
  AnnotationMap& operator=(const AnnotationMap&) = delete;
```

See this for example: https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/profiler/rpc/client/profiler_client.h;l=53

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:206]: Our style guide discourages forward declaration and prefers to include headers directly (include what you use).

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:18]: What is this used for?

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:108]: Double newlines. Use clang format?

File: tensorflow/core/profiler/internal/gpu/BUILD
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/BUILD:148]: The visibility can be restricted to the targets that actually use it. I don't think this needs to be visible globally.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:261]: Could you point me to where api_cb_impl_ and activity_cb_impl_ are being used?

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:363]: (Here and elsewhere) Use tensorflow::profiler::NanosToMicros().

https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/core/profiler/utils/time_utils.h;l=32

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:356]: Prefer to use a std::unique_ptr, so that this will not leak if any other devs changes the (long) code sequence below to return early.

Finally, release() the pointer to swap().

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:362]: Try not to use auto everywhere, use it when type can be inferred within couple lines in the local scope.

**Comment by yisitu** [comment]
@deven-amd  I am unable to test the the functionality on AMD GPUs so you'll have to do that for us.
Since most of the code is in device_tracer_rocm I think the PR is relatively low risk.

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (yisitu) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h:35]: Could you change all the macro arguments to:
1. uppercase?
2. remove the xxx_ prefix?
```suggestion
#define ROCTRACER_API_WRAPPER(API_NAME)					\
```

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (yisitu) [inline tensorflow/stream_executor/rocm/roctracer_wrapper.h:63]: (Similar to an earlier review comment) Replace "xxx_macro" with "DO_FUNC" or "FUNC":
```suggestion
#define FOR_EACH_ROCTRACER_API(DO_FUNC)		\
```

**Review by yisitu — CHANGES_REQUESTED** [review]
(Reviewed and waiting for a round of changes.)

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:37]: this will be fixed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:1]: this will be fixed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:1]: this will be fixed in the next commit/push

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h:1]: this will be fixed in the next commit/push

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h:35]: this will be fixed in the next commit/push

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h:63]: this will be fixed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/BUILD
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/BUILD:148]: the visibility setting is based on the same for the CUDA side, which is also public.

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:18]: this will be removed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:362]: this will be fixed in the next commit/push

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h:65]: this will be fixed in the next commit/push

File: tensorflow/stream_executor/rocm/roctracer_wrapper.h
Comment (deven-amd) [reply tensorflow/stream_executor/rocm/roctracer_wrapper.h:60]: this will be fixed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:261]: https://github.com/tensorflow/tensorflow/pull/47508/files#diff-ed24af3bc5fc7ee922e502256a716aa78ef7ce9cf479a8d56d70d2a147078309R939

https://github.com/tensorflow/tensorflow/pull/47508/files#diff-ed24af3bc5fc7ee922e502256a716aa78ef7ce9cf479a8d56d70d2a147078309R940

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:363]: this will be in the next commit/push

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:55]: this will be fixed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:141]: this will be fixed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:289]: this will be fixed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:180]: this will be fixed in the next commit/push

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:206]: those class implementations are in the `rocm_tracer.cc` file, so no option but to forward declare them.

**Comment by deven-amd** [comment]
@yisitu I have addressed most of your feedback and rebased PR to resolve merge conflicts.

please re-review / re-approve.

@chsigg gentle ping

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:214]: What is commonly done is to declare the class interface in the header, and leave the implementation in the .cc file.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:56]: Could you take care of the refactor in this PR? We are trying not to increase tech debt. This should be a reasonably straightforward change.

File: tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc
Comment (yisitu) [reply tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc]: Hi, I am aware of the intention to pull this into TF 2.5. The quickest way forward is to apply the suggested review comments. Most of these are straightforward refactors. For example, the wrong copyright year might seem inconsequential but we do try to do our part to not raise tech debt. Please unresolve the earlier review feedback that have not been addressed.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.h:214]: > What is commonly done is to declare the class interface in the header, and leave the implementation in the .cc file.

True, but that also adds unnecessary code to the `rocm_tracer.h` file and affects both readability and maintainability. In this case, the two classes that are forward declared, are accesses only within `rocm_tracer.cc` and adding their interface to the `.h` file provides no additional value. I do not see any valid reason for pulling their interface into the header.

>  Our style guide discourages forward declaration and prefers to include headers directly (include what you use).

agree with you on the style guideline, but I am hoping I have made a case here for an exception.

**Comment by deven-amd** [comment]
> Hi, I am aware of the intention to pull this into TF 2.5. The quickest way forward is to apply the suggested review comments. Most of these are straightforward refactors. For example, the wrong copyright year might seem inconsequential but we do try to do our part to not raise tech debt. Please unresolve the earlier review feedback that have not been addressed.

@yisitu , m guessing you were reviewing an outdated version of the changes when you left this comment. 
I did update the copyright year in my previous commits (https://github.com/tensorflow/tensorflow/pull/47508/commits/8625366d5b0d8a93257f48ea177914e5296b14d4) , and only marked as resolved, all the comments that I did address in my previous commits (https://github.com/tensorflow/tensorflow/pull/47508/commits/090e71f42875d09dc51dea6a7b4fbe2e47d27885, https://github.com/tensorflow/tensorflow/pull/47508/commits/8625366d5b0d8a93257f48ea177914e5296b14d4).

**Comment by deven-amd** [comment]
@yisitu I have addressed all of your PR feedback. please re-review.

@chsigg gentle ping

**Comment by deven-amd** [comment]
@yisitu @chsigg gentle ping

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:19]:
```suggestion

```
I think iostream is not needed. Probably left behind from debugging.

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:88]:
```suggestion
std::string GetActivityDomainOpName(uint32_t domain, uint32_t op) {
```

One recent development is that we're qualifying std std::string everywhere. :)

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:461]:
```suggestion
  // TODO(rocm): figure out a way to properly populate this field.
```

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.h
Comment (yisitu) [inline tensorflow/core/profiler/internal/gpu/rocm_tracer.h:213]:
```suggestion

```


Avoid forward declarations.
https://google.github.io/styleguide/cppguide.html#Forward_Declarations

class RocmTracer is in the same file, so this can be fixed by reordering the declaration. In addition, this RocmTracer doesn't seem to be necessary.

**Review by yisitu — APPROVED** [review]
Approved for merge conditioned that the remaining review feedback have been addressed.

**Comment by yisitu** [comment]
@deven-amd 

12:20 PM
From Presubmit:Lint:
> CheckLint/0: Warning (//depot/google3/third_party/tensorflow/core/profiler/METADATA:5:3)
> CheckLint found errors.
> <chrono> is an unapproved C++11 header.  [build/c++11] [5]
> 	//depot/google3/third_party/tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:18
> <thread> is an unapproved C++11 header.  [build/c++11] [5]
> 	//depot/google3/third_party/tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:21
> Redundant blank line at the end of a code block should be deleted.  [whitespace/blank_line] [3]
> 	//depot/google3/third_party/tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:436
> Tab found; better to use spaces  [whitespace/tab] [1]
> 	//depot/google3/third_party/tensorflow/core/profiler/internal/gpu/device_tracer_rocm.cc:722

These are blocking the submit. Please fix. 
Please replace <thread> and <chrono> with absl::SleepFor().

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc:19]: removed in the latest commit

File: tensorflow/core/profiler/internal/gpu/rocm_tracer.cc
Comment (deven-amd) [reply tensorflow/core/profiler/internal/gpu/rocm_tracer.cc]: changed to `std::string` in the latest commit

**Comment by deven-amd** [comment]
@yisitu I have made the changes you have suggested. 

Please re-approve and re-submit.

thanks again

deven

**Review by yisitu — APPROVED** [review]
Approving changes required by linter.

**Comment by deven-amd** [comment]
@gbaned @yisitu PR seems to have gotten stuck in the merge pipeline...anything I can do on my end to move it along? thanks

**Comment by akuegel** [comment]
> @gbaned @yisitu PR seems to have gotten stuck in the merge pipeline...anything I can do on my end to move it along? thanks

There were some internal changes needed to make it work, I took care of it. Still waiting for approvals though, but hopefully it will be merged tomorrow.

**Comment by deven-amd** [comment]
@yisitu @akuegel @gbaned 

thank you all very much for your help in getting this PR merged, it is much appreciated.

**Comment by yisitu** [comment]
@deven-amd @goldiegadde @geetachavan1 

Devan, if you need this in TF 2.5 you might have to request for a cherry pick. I'll let you drive from here.

**Pull request merged**
